File: Java-Civitas/TituloPropiedad.java
package civitas;

import java.util.ArrayList;

public class TituloPropiedad {
    
    private float alquilerBase;
    final private float factorInteresesHipoteca = 1.1f;
    private float factorRevalorizacion;
    private float hipotecaBase;
    private boolean hipotecado;
    private String nombre;
    private int numCasas;
    private int numHoteles;
    private float precioCompra;
    private float precioEdificar;
    private Jugador propietario;
    
    public TituloPropiedad(String nomb, float prec_alquiler, float fac_reval, float prec_hipo, float prec_compra, float prec_edif) {
        nombre = nomb;
        alquilerBase = prec_alquiler;
        factorRevalorizacion = fac_reval;
        hipotecaBase = prec_hipo;
        precioCompra = prec_compra;
        precioEdificar = prec_edif;
        propietario = null;
    }
    
    void actualizaPropietarioPorConversion(Jugador jug) { 
        propietario = jug;
    }
    
    boolean cancelarHipoteca(Jugador jug) {
        throw new UnsupportedOperationException("No implementado");
    }
    
    boolean hipotecar(Jugador jug) {
        throw new UnsupportedOperationException("No implementado");
    }
    
    boolean comprar(Jugador jug) {
        throw new UnsupportedOperationException("No implementado");
    }
    
    boolean construirCasa(Jugador jug) {
        throw new UnsupportedOperationException("No implementado");
    }
    
    boolean construirHotel(Jugador jug) {
        throw new UnsupportedOperationException("No implementado");
    }
    
    private float getImporteHipoteca() {
        return hipotecaBase;
    }
    
    String getNombre() {
        return nombre;
    }
    
    int getNumCasas() {
        return numCasas;
    }
    
    int getNumHoteles() {
        return numHoteles;
    }
    
    float getPrecioCompra() {
        return precioCompra;
    }
    
    float getPrecioEdificar() {
        return precioEdificar;
    }
    
    Jugador getPropietario() {
        return propietario;
    }
    
    public boolean getHipotecado() {
        return hipotecado;
    }
    
    private boolean esEsteElPropietario(Jugador jug) {
        boolean esPropietario = false;
        
        if(getPropietario() == jug)
            esPropietario = true;
        
        return esPropietario;
    }
    
    private float getPrecioAlquiler() {
        float precio_alquiler = 0;
        
        if(!getHipotecado() && !propietarioEncarcelado())
            precio_alquiler = alquilerBase * (1 + (getNumCasas() * 0.5f) + (getNumHoteles() * 2.5f));
        
        return precio_alquiler;
    }
    
    float getImporteCancelarHipoteca() {
        float CantidadRecibida = getImporteHipoteca() * (1 + (getNumCasas() * 0.5f) + (getNumHoteles() * 2.5f));
        return CantidadRecibida * factorInteresesHipoteca;
    }
    
    private boolean propietarioEncarcelado() {
        boolean encarcelado = false;
        
        if(tienePropietario() && propietario.isEncarcelado())
            encarcelado = true;
        
        return encarcelado;
            
    }
    
    void tramitarAlquiler(Jugador jug) {
        float alquiler;
        if(tienePropietario() && !esEsteElPropietario(jug)) {
            alquiler = pagaAlquiler(jug);
            propietario.recibe(getPrecioAlquiler());
        }
    }
    
    boolean tienePropietario() {
        boolean tiene_prop = false;
        
        if(propietario.nombre != "\0")
            tiene_prop = true;
        
        return tiene_prop;
    } 
    
    int cantidadCasasHoteles() {
        return getNumCasas() + getNumHoteles();
    }
    
    private float getPrecioVenta() {
        return getPrecioCompra() + getPrecioEdificar() * factorRevalorizacion;
    }
    
    boolean derruirCasas(int n, Jugador jug) {
        boolean derruir = false;
        if(esEsteElPropietario(jug) && getNumCasas() >= n){
            numCasas -= n;
            derruir = true;
        }
        
        return derruir;
    }
    
    boolean vender(Jugador jug) {
        boolean vendido = false;
        
        if(esEsteElPropietario(jug) && !getHipotecado()){
            propietario.recibe(getPrecioVenta());
            propietario = null;
            numCasas = 0;
            numHoteles = 0;
            vendido = true;
        }
        
        return vendido;
    }
    
    @Override
    public String toString() {
        return "Nombre Propiedad: " + getNombre() + "\n"
                + "Precio Alquiler: " + getPrecioAlquiler() + "\n"
                + "Precio Compra: " + getPrecioCompra() + "\n"
                + "Precio Edificacion: " + getPrecioEdificar() + "\n"
                + "Precio Venta: " + getPrecioVenta() + "\n"
                + "Propietario: " +  getPropietario() + "\n"
                + "Hipoteca: " + getImporteHipoteca() + "\n"
                + "Hipotecada: " + getHipotecado() + "\n"
                + "Numero Casas: " + getNumCasas() + "\n"
                + "Numero Hoteles: " + getNumHoteles();
    }
}

File: Ruby-Civitas/titulo_propiedad.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

module Civitas
  class TituloPropiedad
    
    attr_accessor :nombre, :numCasas, :numHoteles, :precioCompra, :precioEdificar, :propietario, :hipotecado, :hipotecaBase
    
    def initialize(nomb, prec_alquiler, fac_reval, prec_hipo, prec_compra, prec_edif)
      @nombre = nomb
      @alquilerBase = prec_alquiler
      @factorRevalorizacion = fac_reval
      @FACTOR_INTERESES_HIPOTECA = 1.1
      @hipotecaBase = prec_hipo
      @precioCompra = prec_compra
      @precioEdificar = prec_edif
      @propietario = null
      @numCasas = 0
      @numHoteles = 0
      @hipotecado = false
    end
    
    def actualizaPropietarioPorConversion(jug)
      @propietario = jug
    end
    
    def esEsteElPropietario(jug)
      esPropietario = false
      
      if(@propietario == jug)
        esPropietario = true
      end
      return esPropietario
    end
    
    def getPrecioAlquiler()
      precio_alquiler = 0
      
      if(!@hipotecado) && (!@propietario)
        precio_alquiler = @alquilerBase * (1 + (@numCasas * 0.5) + (@numHoteles * 2.5))
      end
      
      return precio_alquiler
    end
    
    def getImporteCancelarHipoteca()
      cantidadRecibida = @hipotecaBase * (1 + (@numCasas * 0.5) + (@numHoteles * 2.5))
      return cantidadRecibida * @FACTOR_INTERESES_HIPOTECA
    end
    
    def propietarioEncarcelado()
      @encarcelado = false
      
      if(tienePropietario()) && (@propietario.isEncarcelado())
        @encarcelado = true
      end
      
      return @encarcelado
    end
    
    def tramitarAlquiler(jug)
      if(tienePropietario()) && (!esEsteElPropietario(jug))
        jug.pagaAlquiler(getPrecioAlquiler())
        propietario.recibe(getPrecioAlquiler())
      end
    end
    
    def tienePropietario()
      tiene_prop = false
      
      if(propietario != nil)
        tiene_prop = true
      end
      
      return tiene_prop
    end
    
    def cantidadCasasHoteles()
      return @numCasas + @numHoteles
    end
    
    def precioVenta()
      return @precioCompra + @precioEdificar * @factorRevalorizacion
    end
    
    def derruirCasas(n, jug)
      derruir = false
      
      if(esEsteElPropietario(jug)) && (@numCasas >= n)
        @numCasas -= n
        derruir = true
      end
      
      return derruir
    end
    
    def vender(jug)
      vendido = false
      
      if(esEsteElPropietario(jug)) && (!@hipotecado)
        @propietario.recibe(getPrecioVenta())
        @propietario = nil
        @numCasas = 0
        @numHoteles = 0
        vendido = true
      end
      
      return vendido
    end
    
    def toString()
      return "Nombre Propiedad: " + @nombre.to_s + "\n"
      + "Precio Alquiler: " + getPrecioAlquiler().to_s + "\n"
      + "Precio Compra: " + @precioCompra.to_s + "\n"
      + "Precio Edificacion: " + @precioEdificar.to_s + "\n"
      + "Precio Venta: " + getPrecioVenta().to_s + "\n"
      + "Propietario: " + @propietario.nombre + "\n"
      + "Hipoteca: " + @hipotecaBase.to_s + "\n"
      + "Hipotecada: " + @hipotecado.to_s + "\n"
      + "Numero Casas: " + @numCasas + "\n"
      + "Numero Hoteles: " + @numHoteles;
    end
  end
end
